clang-tools-extra/clang-tidy/daedalean/PreprocessingDirectivesCheck.cpp

```diff
@@ -0,0 +1,172 @@
+        pragmaDeclaration += data[i];
+      }
+
+      if (pragmaDeclaration == "#pragma once") {
+        if (PragmaOnce[fileId]) {
+          Check.diag(Loc, "Pragma once must be specified only once");
+        }
+
+        PragmaOnce[fileId] = true;
+        return;
+      }
+    }
+
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void MacroDefined(const Token &MacroNameTok,
+                            const MacroDirective *MD) override {
+    if (SM.isWrittenInBuiltinFile(MD->getLocation()) ||
+        SM.isWrittenInCommandLineFile(MD->getLocation()))
+      return;
+
+    Check.diag(MacroNameTok.getLocation(),
+               "Preprocessor directives must not be used");
+  }
+
+  virtual void MacroUndefined(const Token &MacroNameTok,
+                              const MacroDefinition &MD,
+                              const MacroDirective *Undef) override {
+    Check.diag(Undef->getLocation(),
+               "Preprocessor directives must not be used");
+  }
+
+  virtual void If(SourceLocation Loc, SourceRange ConditionRange,
+                  ConditionValueKind ConditionValue) {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Ifdef(SourceLocation Loc, const Token &MacroNameTok,
+                     const MacroDefinition &MD) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Ifndef(SourceLocation Loc, const Token &MacroNameTok,
+                      const MacroDefinition &MD) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Else(SourceLocation Loc, SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Endif(SourceLocation Loc, SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elif(SourceLocation Loc, SourceRange ConditionRange,
+                    ConditionValueKind ConditionValue,
+                    SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elifdef(SourceLocation Loc, const Token &MacroNameTok,
+                       const MacroDefinition &MD) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elifdef(SourceLocation Loc, SourceRange ConditionRange,
+                       SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elifndef(SourceLocation Loc, const Token &MacroNameTok,
+                        const MacroDefinition &MD) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
+
+  virtual void Elifndef(SourceLocation Loc, SourceRange ConditionRange,
+                        SourceLocation IfLoc) override {
+    Check.diag(Loc, "Preprocessor directives must not be used");
+  }
```

Comment:
Should we also catch `#error` and `#line`?

Reply:
Is there any way to use #error and not break compilation without if/ifdef/ifndef ?
For #line  T10253